### PR TITLE
Allow tree-ish versions for ansible-galaxy

### PR DESCRIPTION
**Pull request by willthames**
Rebase of #10620, retaining credit for @semifocused's work. For justification etc., see #10620.

### CODE REVIEW

**Comment by semifocused** [comment]
Very helpful @willthames.  Many thanks for taking on the rebase and looks perfect.

File: lib/ansible/playbook/role/requirement.py [hunk 1]
```diff
@@ -190,6 +190,17 @@ def scm_archive_role(src, scm='git', name=None, version='HEAD'):
...
 
+        if scm == 'git' and version:
+            checkout_cmd = [scm, 'checkout', version]
+            with open('/dev/null', 'w') as devnull:
+                try:
+                    popen = subprocess.Popen(checkout_cmd, cwd=os.path.join(tempdir, name), stdout=devnull, stderr=devnull)
+                except (IOError, OSError):
+                    raise AnsibleError("error executing: %s" % " ".join(checkout_cmd))
+                rc = popen.wait()
+            if rc != 0:
```
Comment (Lujeni) [inline lib/ansible/playbook/role/requirement.py:201]: There is a reason do not use a simple test?
imho, it's more zen :)

``` python
if rc:
   ...
```

File: lib/ansible/playbook/role/requirement.py [hunk 2]
```diff
@@ -190,6 +190,17 @@ def scm_archive_role(src, scm='git', name=None, version='HEAD'):
...
+                raise AnsibleError("- command %s failed in directory %s (rc=%s)" % (' '.join(checkout_cmd), tempdir, rc))
```
Comment (bcoca) [inline lib/ansible/playbook/role/requirement.py:202]: it wold be helpful to get the error from the scm, but i would worry then that the wait call might block, it would require changing the code to account for this.

File: lib/ansible/galaxy/role.py [hunk 3]
```diff
@@ -310,5 +310,3 @@ def spec(self):
         }
         """
         return dict(scm=self.scm, src=self.src, version=self.version, name=self.name)
-
-
```
Comment (bcoca) [inline lib/ansible/galaxy/role.py:314]: not sure why this file is changed

File: lib/ansible/playbook/role/requirement.py (hunk 1 above)
Comment (semifocused) [inline lib/ansible/playbook/role/requirement.py:201]: The approach in the modification regarding testing `rc` was "adopt-and-go" with existing coding conventions.  Shifting those conventions for enhanced zen may be desired, but the scope of doing so would be larger than this PR.  Considering occurences in the ansible code base of `rc != 0`:

```
~/git/ansible$ find . -type f | xargs -I FILE grep "rc != 0" FILE /dev/null | wc -l
353
```

Making this change may or may not be desired, if so might it best be undertaken in its own PR?

Comment (bcoca) [inline lib/ansible/playbook/role/requirement.py:201]: i'm fine with the less 'zenliness' of `rc != 0` as this reflects the non pythonic shelling out to a command line utility

File: lib/ansible/galaxy/role.py (hunk 3 above)
Comment (willthames) [inline lib/ansible/galaxy/role.py:314]: Not sure why there are a couple of empty lines at the bottom of the file either :). Probably don't need this though.

File: lib/ansible/playbook/role/requirement.py (hunk 2 above)
Comment (willthames) [inline lib/ansible/playbook/role/requirement.py:202]: Very similar (to the extent it might be worth abstracting it) code is in `role/requirement.py` in `scm_archive_role` twice. If it's good enough for `git clone` and `git archive`, surely it's good enough for `git checkout`.

**Comment by willthames** [comment]
Please can you be very explicit about what is blocking the merge of this PR. I'm not convinced it `needs_revision` but maybe that's because we're fighting a review -> update -> review -> update -> review cycle that we've been around a dozen times and it would be nice to actually get this feature merged at last.

File: lib/ansible/playbook/role/requirement.py (hunk 2 above)
Comment (bcoca) [inline lib/ansible/playbook/role/requirement.py:202]: i've been meaning to fix that also, i finally let it go cause I did not have the time, i spent too much trying to restructure galaxy and roles. This is not a showstopper.

**Comment by semifocused** [comment]
@bcoca ping on merge status?

**Comment by bcoca** [comment]
i'll get back to this once we finish releasing 2.0, which should be very soon

**Comment by semifocused** [comment]
Happy New Year @bcoca @willthames!!

Any further update on ability / willingness to get this merged?

**Comment by bcoca** [comment]
still working on 2.0, will get to this probably next week

**Comment by willthames** [comment]
@bcoca, can we finally have this included now that 2.0 has gone live?

**Comment by scottcunningham** [comment]
Any updates on this? I'd love to be able to use it.

**Comment by jimi-c** [comment]
Merged, thanks!
